subsys/fs/littlefs_fs.c

```diff
@@ -619,40 +582,123 @@ static lfs_size_t get_block_size(const struct flash_area *fa)
 	return ctx.max_size;
 }
 
-static int littlefs_flash_init(struct fs_mount_t *mountp)
+static int littlefs_init_fmp(struct fs_mount_t *mountp)
 {
 	unsigned int area_id = (uintptr_t)mountp->storage_dev;
 	struct fs_littlefs *fs = mountp->fs_data;
-	const struct flash_area **fap = (const struct flash_area **)&fs->backend;
+	const struct flash_area *area;
 	const struct device *dev;
 	int ret;
 
 	/* Open flash area */
-	ret = flash_area_open(area_id, fap);
-	if ((ret < 0) || (*fap == NULL)) {
+	ret = flash_area_open(area_id, &area);
+	if ((ret < 0) || (area == NULL)) {
 		LOG_ERR("can't open flash area %d", area_id);
 		return -ENODEV;
 	}
 
 	LOG_DBG("FS area %u at 0x%x for %u bytes", area_id,
-		(uint32_t)(*fap)->fa_off, (uint32_t)(*fap)->fa_size);
+		(uint32_t)(area->fa_off), (uint32_t)(area->fa_size));
 
-	dev = flash_area_get_device(*fap);
+	dev = flash_area_get_device(area);
 	if (dev == NULL) {
 		LOG_ERR("can't get flash device: %s",
-			log_strdup((*fap)->fa_dev_name));
+			log_strdup(area->fa_dev_name));
 		return -ENODEV;
 	}
 
-	fs->backend = (void *) *fap;
+	fs->backend = (void *)area;
 	return 0;
 }
 
+/* Block device (blk) backend functions */
+static inline bool lfs_on_blkdev(struct fs_mount_t *mountp)
+{
+	return mountp->flags & FS_MOUNT_FLAG_USE_DISK_ACCESS;
+}
+static int lfs_api_read_blk(const struct lfs_config *c, lfs_block_t block,
+			    lfs_off_t off, void *buffer, lfs_size_t size)
+{
+	const char *disk = c->context;
+	int rc = disk_access_read(disk, buffer, block,
+				  size / c->block_size);
+
+	return errno_to_lfs(rc);
+}
+
+static int lfs_api_prog_blk(const struct lfs_config *c, lfs_block_t block,
+			    lfs_off_t off, const void *buffer, lfs_size_t size)
+{
+	const char *disk = c->context;
+	int rc = disk_access_write(disk, buffer, block, size / c->block_size);
+
+	return errno_to_lfs(rc);
+}
+
+static int lfs_api_erase_blk(const struct lfs_config *c, lfs_block_t block)
+{
+	return 0;
+}
+
+static int lfs_api_sync_blk(const struct lfs_config *c)
+{
+	const char *disk = c->context;
+	int rc = disk_access_ioctl(disk, DISK_IOCTL_CTRL_SYNC, NULL);
+
+	return errno_to_lfs(rc);
+}
+
+static int littlefs_init_blk(struct fs_mount_t *mountp)
+{
+	struct fs_littlefs *fs = mountp->fs_data;
+	int ret;
+
+	ret = disk_access_init((char *)mountp->storage_dev);
+	if (ret < 0) {
+		LOG_ERR("Storage init ERROR!");
+		goto end;
```

Comment:
return ret; and remove the `end` label

Reply:
This is a general flow I follow, minimise the number of exit points. The general flow would be:
a. At the start of a routine check input arguments and return quickly,
b. During execution check for errors and jump to an exit point if failure (reset things if required)
c. If succesful during execution return 0 or flow to the end if already there.
This flow is very easy to understand and to follow.

Reply:
> This flow is very easy to understand and to follow.

But OK, if this is Zephyr's accepted paradigm, then I don't mind.